libjsconsole/JSConsole.cpp

```diff
@@ -0,0 +1,85 @@
+/*
+	This file is part of cpp-ethereum.
+
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file JSConsole.cpp
+ * @author Marek Kotewicz <marek@ethdev.com>
+ * @date 2015
+ * Ethereum client.
+ */
+
+#include <iostream>
+#include <libdevcore/Log.h>
+#include "JSConsole.h"
+#include "JSV8Connector.h"
+#include "libjsconsole/JSConsoleResources.hpp"
+
+// TODO! make readline optional!
+#include <readline/readline.h>
+#include <readline/history.h>
+
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+
+JSConsole::JSConsole(WebThreeDirect& _web3, std::vector<dev::KeyPair> const& _accounts):
+		m_engine(),
+		m_printer(m_engine)
+{
+	m_jsonrpcConnector.reset(new JSV8Connector(m_engine));
+	m_jsonrpcServer.reset(new WebThreeStubServer(*m_jsonrpcConnector.get(), _web3, _accounts));
+}
+
+void JSConsole::repl() const
+{
+	string cmd = "";
+	g_logPost = [](std::string const& a, char const*) { cout << "\r           \r" << a << endl << flush; rl_forced_update_display(); };
+
+	bool isEmpty = true;
+	int openBrackets = 0;
+	do {
+		char* buff = readline(promptForIndentionLevel(openBrackets).c_str());
+		isEmpty = !(buff && *buff);
+		if (!isEmpty)
+		{
+			cmd += string(buff);
+			cmd += " ";
+			free(buff);
+			int open = count(cmd.begin(), cmd.end(), '{');
```

Comment:
>This counting can match brackets of different type i.e. "func(}" and does not account for quoted brackets.

You are right, I haven't though about this. Can we make a separate issue for this? It think it's not a blocker, cause it will it's important only in few cases.

>Is it possible to just try and compile a script at this point and make a decision to get more input based on compilation error?

I'm afraid that it won't be so easy... Some nice regexp might be better.